src/citrine/_rest/collection.py

```diff
@@ -222,3 +228,60 @@ def _check_experimental(self, data):
                 "\n  ".join(data.get("experimental_reasons") or ["Unknown reason"])
             )
             warnings.warn(msg)
+
+    def _poll_for_job_completion(self, project_id: UUID,
+                                 job: Union[JobSubmissionResponse, UUID], *,
+                                 timeout: float = 2 * 60,
+                                 polling_delay: float = 2.0) -> JobStatusResponse:
+        """
+        Polls for job completion given a timeout, failing with an exception on job failure.
+
+        This polls for job completion given the Job ID, failing appropriately if the job result
+        was not successful.
+
+        Parameters
+        ----------
+        job
+            The job submission object or job ID that was given from a job submission.
+        timeout
+            Amount of time to wait on the job (in seconds) before giving up. Defaults
+            to 2 minutes. Note that this number has no effect on the underlying job
+            itself, which can also time out server-side.
+
+        Returns
+        -------
+        JobStatusResponse
+            The job response information that can be used to extract relevant
+            information from the completed job.
+
+        """
+        if isinstance(job, JobSubmissionResponse):
+            job_id = job.job_id
+        else:
+            job_id = job  # pragma: no cover
+        path = 'projects/{}/execution/job-status'.format(project_id)
+        params = {'job_id': job_id}
+        start_time = time()
+        while True:
+            response = self.session.get_resource(path=path, params=params)
+            status: JobStatusResponse = JobStatusResponse.build(response)
+            if status.status in ['Success', 'Failure']:
+                break
+            elif time() - start_time < timeout:
+                logger.info('Job still in progress, polling status again in {:.2f} seconds.'
+                            .format(polling_delay))
+
+                sleep(polling_delay)
+            else:
+                logger.error('Job exceeded user timeout of {} seconds.'.format(timeout))
+                logger.debug('Last status: {}'.format(status.dump()))
+                raise TimeoutError('Job {} timed out.'.format(job_id))
```

Comment:
Could we raise a custom exception here (derived from `TimeoutError`), e.g. [these](https://github.com/CitrineInformatics/citrine-python/blob/master/src/citrine/exceptions.py).

Reply:
Done - we now raise a custom PollingTimeoutError.